<!DOCTYPE html>

<html lang="en">
<head>
	<meta charset="utf-8">
	<meta name="viewport" content="width=device-width">
	<title>Bioinfo functions Global</title>

	<!--[if lt IE 9]>
	<script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
	<![endif]-->
	<link type="text/css" rel="stylesheet" href="styles/sunlight.default.css">

	<link type="text/css" rel="stylesheet" href="styles/site.yeti.css">

</head>

<body>

<div class="navbar navbar-default navbar-fixed-top navbar-inverse">
<div class="container">
	<div class="navbar-header">
		<a class="navbar-brand" href="index.html">Bioinfo functions</a>
		<button class="navbar-toggle" type="button" data-toggle="collapse" data-target="#topNavigation">
			<span class="icon-bar"></span>
			<span class="icon-bar"></span>
			<span class="icon-bar"></span>
        </button>
	</div>
	<div class="navbar-collapse collapse" id="topNavigation">
		<ul class="nav navbar-nav">
			
			<li class="dropdown">
				<a href="global.html" class="dropdown-toggle" data-toggle="dropdown">Global<b class="caret"></b></a>
				<ul class="dropdown-menu ">
					<li><a href="global.html#clean_peptide">clean_peptide</a></li><li><a href="global.html#complement">complement</a></li><li><a href="global.html#convertDegenerates">convertDegenerates</a></li><li><a href="global.html#dna_mod">dna_mod</a></li><li><a href="global.html#dnaMw">dnaMw</a></li><li><a href="global.html#dnaStats">dnaStats</a></li><li><a href="global.html#MolarExtinction">MolarExtinction</a></li><li><a href="global.html#Mutation2Codon">Mutation2Codon</a></li><li><a href="global.html#Mutation2Seq">Mutation2Seq</a></li><li><a href="global.html#pairwiseAlignDna">pairwiseAlignDna</a></li><li><a href="global.html#pairwiseAlignProtein">pairwiseAlignProtein</a></li><li><a href="global.html#pcrPrimerStats">pcrPrimerStats</a></li><li><a href="global.html#peptide_Cmod">peptide_Cmod</a></li><li><a href="global.html#peptide_intern_mod">peptide_intern_mod</a></li><li><a href="global.html#peptide_Nmod">peptide_Nmod</a></li><li><a href="global.html#protein_mw">protein_mw</a></li><li><a href="global.html#proteinGravy">proteinGravy</a></li><li><a href="global.html#proteinIep">proteinIep</a></li><li><a href="global.html#proteinStats">proteinStats</a></li><li><a href="global.html#restSummary">restSummary</a></li><li><a href="global.html#revComp">revComp</a></li><li><a href="global.html#reverse">reverse</a></li><li><a href="global.html#Seq1">Seq1</a></li><li><a href="global.html#Seq3">Seq3</a></li><li><a href="global.html#SeqDiff">SeqDiff</a></li><li><a href="global.html#translate">translate</a></li>
				</ul>
			</li>
			
		</ul>
        
	</div>

</div>
</div>


<div class="container" id="toc-content">
<div class="row">

	
	<div class="col-md-8">
	
		<div id="main">
			

	<h1 class="page-title">Global</h1>
<section>

<header>
    
        <h2>
        
        </h2>
        
    
</header>


<article>
    <div class="container-overview">
    
        

        
<dl class="details">
    

	

	

	

    

    

    

    

	

	

	

	

	



	

	

	

	
</dl>


        
    
    </div>

    

    

    

    

    

    

    
        <h3 class="subsection-title">Methods</h3>

        <dl>
            
<hr>
<dt>
    <h4 class="name" id="clean_peptide"><span class="type-signature"></span>clean_peptide(peptideSequence)</h4>
    
    
</dt>
<dd>

    
    <div class="description">
        <p>Function to analyze and return clean peptide sequence; in the sequence, modification is enclosed in ()</p>
    </div>
    

    

    

    
    
        <h5>Parameters:</h5>
        

<table class="params table table-striped">
    <thead>
    <tr>
        
        <th>Name</th>
        

        <th>Type</th>

        

        

        <th class="last">Description</th>
    </tr>
    </thead>

    <tbody>
    

        <tr>
            
                <td class="name"><code>peptideSequence</code></td>
            

            <td class="type">
            
                
<span class="param-type">string</span>



            
            </td>

            

            

            <td class="description last"><p>Input Protein one letter sequence</p></td>
        </tr>

    
    </tbody>
</table>

    

    
<dl class="details">
    

	

	

	

    

    

    

    

	

	

	

	

	



	

	

	

	
</dl>


    

    

    

    

    

    
    <h5>Returns:</h5>
    
            
<div class="param-desc">
    <p>clean peptide sequence</p>
</div>




        

    
</dd>

        
            
<hr>
<dt>
    <h4 class="name" id="complement"><span class="type-signature"></span>complement(dnaSequence)</h4>
    
    
</dt>
<dd>

    
    <div class="description">
        <p>Function to complement counterpart of input DNA sequence.<br>
<a href="http://www.bioinformatics.org/sms2/rev_comp.html">http://www.bioinformatics.org/sms2/rev_comp.html</a></p>
    </div>
    

    

    

    
    
        <h5>Parameters:</h5>
        

<table class="params table table-striped">
    <thead>
    <tr>
        
        <th>Name</th>
        

        <th>Type</th>

        

        

        <th class="last">Description</th>
    </tr>
    </thead>

    <tbody>
    

        <tr>
            
                <td class="name"><code>dnaSequence</code></td>
            

            <td class="type">
            
                
<span class="param-type">string</span>



            
            </td>

            

            

            <td class="description last"><p>Input DNA sequence</p></td>
        </tr>

    
    </tbody>
</table>

    

    
<dl class="details">
    

	

	

	

    

    

    

    

	

	

	

	

	



	

	

	

	
</dl>


    

    

    

    

    

    
    <h5>Returns:</h5>
    
            
<div class="param-desc">
    <p>complement of input DNA sequence</p>
</div>




        

    
</dd>

        
            
<hr>
<dt>
    <h4 class="name" id="convertDegenerates"><span class="type-signature"></span>convertDegenerates(sequence)</h4>
    
    
</dt>
<dd>

    
    <div class="description">
        <p>Function to convert degenerate base in a sequence.<br>
<a href="https://www.bioinformatics.org/sms/iupac.html">https://www.bioinformatics.org/sms/iupac.html</a></p>
    </div>
    

    

    

    
    
        <h5>Parameters:</h5>
        

<table class="params table table-striped">
    <thead>
    <tr>
        
        <th>Name</th>
        

        <th>Type</th>

        

        

        <th class="last">Description</th>
    </tr>
    </thead>

    <tbody>
    

        <tr>
            
                <td class="name"><code>sequence</code></td>
            

            <td class="type">
            
                
<span class="param-type">string</span>



            
            </td>

            

            

            <td class="description last"><p>Input DNA sequence</p></td>
        </tr>

    
    </tbody>
</table>

    

    
<dl class="details">
    

	

	

	

    

    

    

    

	

	

	

	

	



	

	

	

	
</dl>


    

    

    

    

    

    
    <h5>Returns:</h5>
    
            
<div class="param-desc">
    <p>degenerate base in a sequence</p>
</div>




        

    
</dd>

        
            
<hr>
<dt>
    <h4 class="name" id="dna_mod"><span class="type-signature"></span>dna_mod(dnaSequence)</h4>
    
    
</dt>
<dd>

    
    <div class="description">
        <p>Function to analyze and return modification list of a dna sequence; in the sequence, dna modification is enclosed in //</p>
    </div>
    

    

    

    
    
        <h5>Parameters:</h5>
        

<table class="params table table-striped">
    <thead>
    <tr>
        
        <th>Name</th>
        

        <th>Type</th>

        

        

        <th class="last">Description</th>
    </tr>
    </thead>

    <tbody>
    

        <tr>
            
                <td class="name"><code>dnaSequence</code></td>
            

            <td class="type">
            
                
<span class="param-type">string</span>



            
            </td>

            

            

            <td class="description last"><p>Input DNA one letter sequence</p></td>
        </tr>

    
    </tbody>
</table>

    

    
<dl class="details">
    

	

	

	

    

    

    

    

	

	

	

	

	



	

	

	

	
</dl>


    

    

    

    

    

    
    <h5>Returns:</h5>
    
            


<dl>
    <dt>
        Type
    </dt>
    <dd>
        
<span class="param-type">mods</span>



    </dd>
</dl>


        

    
</dd>

        
            
<hr>
<dt>
    <h4 class="name" id="dnaMw"><span class="type-signature"></span>dnaMw(sequence, strandType, topology)</h4>
    
    
</dt>
<dd>

    
    <div class="description">
        <p>Function to calculate calculates molecular weight (Da) of either single stranded or double stranded DNA.<br>
<a href="http://www.bioinformatics.org/sms2/dna_mw.html">http://www.bioinformatics.org/sms2/dna_mw.html</a></p>
    </div>
    

    

    

    
    
        <h5>Parameters:</h5>
        

<table class="params table table-striped">
    <thead>
    <tr>
        
        <th>Name</th>
        

        <th>Type</th>

        

        
        <th>Default</th>
        

        <th class="last">Description</th>
    </tr>
    </thead>

    <tbody>
    

        <tr>
            
                <td class="name"><code>sequence</code></td>
            

            <td class="type">
            
                
<span class="param-type">string</span>



            
            </td>

            

            
                <td class="default">
                
                </td>
            

            <td class="description last"><p>Input DNA</p></td>
        </tr>

    

        <tr>
            
                <td class="name"><code>strandType</code></td>
            

            <td class="type">
            
                
<span class="param-type">"double"</span>



            
            </td>

            

            
                <td class="default">
                
                    double
                
                </td>
            

            <td class="description last"><p>double (default) or single</p></td>
        </tr>

    

        <tr>
            
                <td class="name"><code>topology</code></td>
            

            <td class="type">
            
                
<span class="param-type">"linear"</span>



            
            </td>

            

            
                <td class="default">
                
                    linear
                
                </td>
            

            <td class="description last"><p>linear (default) or circular</p></td>
        </tr>

    
    </tbody>
</table>

    

    
<dl class="details">
    

	

	

	

    

    

    

    

	

	

	

	

	



	

	

	

	
</dl>


    

    

    

    

    

    
    <h5>Returns:</h5>
    
            
<div class="param-desc">
    <p>molecular weight of DNA (Da)</p>
</div>




        

    
</dd>

        
            
<hr>
<dt>
    <h4 class="name" id="dnaStats"><span class="type-signature"></span>dnaStats(dnaSequence)</h4>
    
    
</dt>
<dd>

    
    <div class="description">
        <p>Function to DNA stats of a sequence.<br>
<a href="http://www.bioinformatics.org/sms2/dna_stats.html">http://www.bioinformatics.org/sms2/dna_stats.html</a></p>
    </div>
    

    

    

    
    
        <h5>Parameters:</h5>
        

<table class="params table table-striped">
    <thead>
    <tr>
        
        <th>Name</th>
        

        <th>Type</th>

        

        

        <th class="last">Description</th>
    </tr>
    </thead>

    <tbody>
    

        <tr>
            
                <td class="name"><code>dnaSequence</code></td>
            

            <td class="type">
            
                
<span class="param-type">string</span>



            
            </td>

            

            

            <td class="description last"><p>Input DNA sequence</p></td>
        </tr>

    
    </tbody>
</table>

    

    
<dl class="details">
    

	

	

	

    

    

    

    

	

	

	

	

	



	

	

	

	
</dl>


    

    

    

    

    

    
    <h5>Returns:</h5>
    
            
<div class="param-desc">
    <p>NT composition stats of Input DNA sequence</p>
</div>




        

    
</dd>

        
            
<hr>
<dt>
    <h4 class="name" id="MolarExtinction"><span class="type-signature"></span>MolarExtinction(proteinSequence, Cys)</h4>
    
    
</dt>
<dd>

    
    <div class="description">
        <p>Function to calculate Calculate the molar extinction coefficient (units of  M-1 cm-1, at 280 nm measured in water). <a href="https://web.expasy.org/cgi-bin/protparam/protparam">https://web.expasy.org/cgi-bin/protparam/protparam</a></p>
    </div>
    

    

    

    
    
        <h5>Parameters:</h5>
        

<table class="params table table-striped">
    <thead>
    <tr>
        
        <th>Name</th>
        

        <th>Type</th>

        

        
        <th>Default</th>
        

        <th class="last">Description</th>
    </tr>
    </thead>

    <tbody>
    

        <tr>
            
                <td class="name"><code>proteinSequence</code></td>
            

            <td class="type">
            
                
<span class="param-type">string</span>



            
            </td>

            

            
                <td class="default">
                
                </td>
            

            <td class="description last"><p>Input Protein one letter sequence</p></td>
        </tr>

    

        <tr>
            
                <td class="name"><code>Cys</code></td>
            

            <td class="type">
            
                
<span class="param-type">"reduced"</span>



            
            </td>

            

            
                <td class="default">
                
                    reduced
                
                </td>
            

            <td class="description last"><p>oxidation state of Cys reduced (default) or disulfide</p></td>
        </tr>

    
    </tbody>
</table>

    

    
<dl class="details">
    

	

	

	

    

    

    

    

	

	

	

	

	



	

	

	

	
</dl>


    

    

    

    

    

    
    <h5>Returns:</h5>
    
            
<div class="param-desc">
    <p>Extiniction Coefficient of input protein</p>
</div>




        

    
</dd>

        
            
<hr>
<dt>
    <h4 class="name" id="Mutation2Codon"><span class="type-signature"></span>Mutation2Codon(dnaSequence, mutationlist)</h4>
    
    
</dt>
<dd>

    
    <div class="description">
        <p>Function to mutate DNA given AA to a codon at specified location</p>
    </div>
    

    

    

    
    
        <h5>Parameters:</h5>
        

<table class="params table table-striped">
    <thead>
    <tr>
        
        <th>Name</th>
        

        <th>Type</th>

        

        

        <th class="last">Description</th>
    </tr>
    </thead>

    <tbody>
    

        <tr>
            
                <td class="name"><code>dnaSequence</code></td>
            

            <td class="type">
            
                
<span class="param-type">string</span>



            
            </td>

            

            

            <td class="description last"><p>Input one letter sequence 1</p></td>
        </tr>

    

        <tr>
            
                <td class="name"><code>mutationlist</code></td>
            

            <td class="type">
            
                
<span class="param-type">"A100ATG"</span>



            
            </td>

            

            

            <td class="description last"><p>Comma separated mutation list in format of A#ATG</p></td>
        </tr>

    
    </tbody>
</table>

    

    
<dl class="details">
    

	

	

	

    

    

    

    

	

	

	

	

	



	

	

	

	
</dl>


    

    

    

    

    

    
    <h5>Returns:</h5>
    
            
<div class="param-desc">
    <p>mutated DNA sequence</p>
</div>




        

    
</dd>

        
            
<hr>
<dt>
    <h4 class="name" id="Mutation2Seq"><span class="type-signature"></span>Mutation2Seq(proteinSequence, mutationlist)</h4>
    
    
</dt>
<dd>

    
    <div class="description">
        <p>Function to mutate protein given AA to AA at specified location</p>
    </div>
    

    

    

    
    
        <h5>Parameters:</h5>
        

<table class="params table table-striped">
    <thead>
    <tr>
        
        <th>Name</th>
        

        <th>Type</th>

        

        

        <th class="last">Description</th>
    </tr>
    </thead>

    <tbody>
    

        <tr>
            
                <td class="name"><code>proteinSequence</code></td>
            

            <td class="type">
            
                
<span class="param-type">string</span>



            
            </td>

            

            

            <td class="description last"><p>Input one letter sequence 1</p></td>
        </tr>

    

        <tr>
            
                <td class="name"><code>mutationlist</code></td>
            

            <td class="type">
            
                
<span class="param-type">"A100M"</span>



            
            </td>

            

            

            <td class="description last"><p>Comma separated mutation list in format of A#X</p></td>
        </tr>

    
    </tbody>
</table>

    

    
<dl class="details">
    

	

	

	

    

    

    

    

	

	

	

	

	



	

	

	

	
</dl>


    

    

    

    

    

    
    <h5>Returns:</h5>
    
            
<div class="param-desc">
    <p>mutated seqeunces</p>
</div>




        

    
</dd>

        
            
<hr>
<dt>
    <h4 class="name" id="pairwiseAlignDna"><span class="type-signature"></span>pairwiseAlignDna(dnaSequenceOne, dnaSequenceTwo, matchScore, mismatchScore, gapPenalty, beginGapPenalty, endGapPenalty)</h4>
    
    
</dt>
<dd>

    
    <div class="description">
        <p>Function to determines the optimal global alignment of two dna sequences.<br>
<a href="http://www.bioinformatics.org/sms2/pairwise_align_dna.html">http://www.bioinformatics.org/sms2/pairwise_align_dna.html</a></p>
    </div>
    

    

    

    
    
        <h5>Parameters:</h5>
        

<table class="params table table-striped">
    <thead>
    <tr>
        
        <th>Name</th>
        

        <th>Type</th>

        

        
        <th>Default</th>
        

        <th class="last">Description</th>
    </tr>
    </thead>

    <tbody>
    

        <tr>
            
                <td class="name"><code>dnaSequenceOne</code></td>
            

            <td class="type">
            
                
<span class="param-type">string</span>



            
            </td>

            

            
                <td class="default">
                
                </td>
            

            <td class="description last"><p>Input DNA One</p></td>
        </tr>

    

        <tr>
            
                <td class="name"><code>dnaSequenceTwo</code></td>
            

            <td class="type">
            
                
<span class="param-type">string</span>



            
            </td>

            

            
                <td class="default">
                
                </td>
            

            <td class="description last"><p>Input DNA Two</p></td>
        </tr>

    

        <tr>
            
                <td class="name"><code>matchScore</code></td>
            

            <td class="type">
            
                
<span class="param-type">matchScore</span>



            
            </td>

            

            
                <td class="default">
                
                    2
                
                </td>
            

            <td class="description last"><p>default 2</p></td>
        </tr>

    

        <tr>
            
                <td class="name"><code>mismatchScore</code></td>
            

            <td class="type">
            
                
<span class="param-type">mismatchScore</span>



            
            </td>

            

            
                <td class="default">
                
                </td>
            

            <td class="description last"><p>default -1</p></td>
        </tr>

    

        <tr>
            
                <td class="name"><code>gapPenalty</code></td>
            

            <td class="type">
            
                
<span class="param-type">gapPenalty</span>



            
            </td>

            

            
                <td class="default">
                
                    2
                
                </td>
            

            <td class="description last"><p>default 2</p></td>
        </tr>

    

        <tr>
            
                <td class="name"><code>beginGapPenalty</code></td>
            

            <td class="type">
            
                
<span class="param-type">beginGapPenalty</span>



            
            </td>

            

            
                <td class="default">
                
                    0
                
                </td>
            

            <td class="description last"><p>default 0</p></td>
        </tr>

    

        <tr>
            
                <td class="name"><code>endGapPenalty</code></td>
            

            <td class="type">
            
                
<span class="param-type">endGapPenalty</span>



            
            </td>

            

            
                <td class="default">
                
                    0
                
                </td>
            

            <td class="description last"><p>default 0</p></td>
        </tr>

    
    </tbody>
</table>

    

    
<dl class="details">
    

	

	

	

    

    

    

    

	

	

	

	

	



	

	

	

	
</dl>


    

    

    

    

    

    
    <h5>Returns:</h5>
    
            
<div class="param-desc">
    <p>alignment of two protein</p>
</div>




        

    
</dd>

        
            
<hr>
<dt>
    <h4 class="name" id="pairwiseAlignProtein"><span class="type-signature"></span>pairwiseAlignProtein(proteinSequenceOne, proteinSequenceTwo, matrix, beginGapPenalty, gapPenalty, endGapPenalty)</h4>
    
    
</dt>
<dd>

    
    <div class="description">
        <p>Function to determines the optimal global alignment of two protein sequences.<br>
<a href="http://www.bioinformatics.org/sms2/pairwise_align_protein.html">http://www.bioinformatics.org/sms2/pairwise_align_protein.html</a></p>
    </div>
    

    

    

    
    
        <h5>Parameters:</h5>
        

<table class="params table table-striped">
    <thead>
    <tr>
        
        <th>Name</th>
        

        <th>Type</th>

        

        
        <th>Default</th>
        

        <th class="last">Description</th>
    </tr>
    </thead>

    <tbody>
    

        <tr>
            
                <td class="name"><code>proteinSequenceOne</code></td>
            

            <td class="type">
            
                
<span class="param-type">string</span>



            
            </td>

            

            
                <td class="default">
                
                </td>
            

            <td class="description last"><p>Input Protein One</p></td>
        </tr>

    

        <tr>
            
                <td class="name"><code>proteinSequenceTwo</code></td>
            

            <td class="type">
            
                
<span class="param-type">string</span>



            
            </td>

            

            
                <td class="default">
                
                </td>
            

            <td class="description last"><p>Input Protein Two</p></td>
        </tr>

    

        <tr>
            
                <td class="name"><code>matrix</code></td>
            

            <td class="type">
            
                
<span class="param-type">"blosum62"</span>



            
            </td>

            

            
                <td class="default">
                
                    blosum62
                
                </td>
            

            <td class="description last"><p>Scoring matrix: blosum62 (default), pam30, pam70, blosum80, blosum45</p></td>
        </tr>

    

        <tr>
            
                <td class="name"><code>beginGapPenalty</code></td>
            

            <td class="type">
            
                
<span class="param-type">0</span>



            
            </td>

            

            
                <td class="default">
                
                    0
                
                </td>
            

            <td class="description last"><p>default 0</p></td>
        </tr>

    

        <tr>
            
                <td class="name"><code>gapPenalty</code></td>
            

            <td class="type">
            
                
<span class="param-type">2</span>



            
            </td>

            

            
                <td class="default">
                
                    2
                
                </td>
            

            <td class="description last"><p>default 2</p></td>
        </tr>

    

        <tr>
            
                <td class="name"><code>endGapPenalty</code></td>
            

            <td class="type">
            
                
<span class="param-type">0</span>



            
            </td>

            

            
                <td class="default">
                
                    0
                
                </td>
            

            <td class="description last"><p>default 0</p></td>
        </tr>

    
    </tbody>
</table>

    

    
<dl class="details">
    

	

	

	

    

    

    

    

	

	

	

	

	



	

	

	

	
</dl>


    

    

    

    

    

    
    <h5>Returns:</h5>
    
            
<div class="param-desc">
    <p>alignment of two protein</p>
</div>




        

    
</dd>

        
            
<hr>
<dt>
    <h4 class="name" id="pcrPrimerStats"><span class="type-signature"></span>pcrPrimerStats(dnaSequence, isPhosphorylated)</h4>
    
    
</dt>
<dd>

    
    <div class="description">
        <p>Function to report describe the properties of a primer, including melting temperature, percent GC content, and PCR suitability.<br>
<a href="http://www.bioinformatics.org/sms2/pcr_primer_stats.html">http://www.bioinformatics.org/sms2/pcr_primer_stats.html</a></p>
    </div>
    

    

    

    
    
        <h5>Parameters:</h5>
        

<table class="params table table-striped">
    <thead>
    <tr>
        
        <th>Name</th>
        

        <th>Type</th>

        

        
        <th>Default</th>
        

        <th class="last">Description</th>
    </tr>
    </thead>

    <tbody>
    

        <tr>
            
                <td class="name"><code>dnaSequence</code></td>
            

            <td class="type">
            
                
<span class="param-type">string</span>



            
            </td>

            

            
                <td class="default">
                
                </td>
            

            <td class="description last"><p>Input DNA sequence</p></td>
        </tr>

    

        <tr>
            
                <td class="name"><code>isPhosphorylated</code></td>
            

            <td class="type">
            
                
<span class="param-type">false</span>



            
            </td>

            

            
                <td class="default">
                
                    false
                
                </td>
            

            <td class="description last"><p>default (false)</p></td>
        </tr>

    
    </tbody>
</table>

    

    
<dl class="details">
    

	

	

	

    

    

    

    

	

	

	

	

	



	

	

	

	
</dl>


    

    

    

    

    

    
    <h5>Returns:</h5>
    
            
<div class="param-desc">
    <p>primer stats</p>
</div>




        

    
</dd>

        
            
<hr>
<dt>
    <h4 class="name" id="peptide_Cmod"><span class="type-signature"></span>peptide_Cmod(peptideSequence)</h4>
    
    
</dt>
<dd>

    
    <div class="description">
        <p>Function to analyze and return C-term mod of peptide sequence; in the sequence, modification is enclosed in ()</p>
    </div>
    

    

    

    
    
        <h5>Parameters:</h5>
        

<table class="params table table-striped">
    <thead>
    <tr>
        
        <th>Name</th>
        

        <th>Type</th>

        

        

        <th class="last">Description</th>
    </tr>
    </thead>

    <tbody>
    

        <tr>
            
                <td class="name"><code>peptideSequence</code></td>
            

            <td class="type">
            
                
<span class="param-type">string</span>



            
            </td>

            

            

            <td class="description last"><p>Input Protein one letter sequence</p></td>
        </tr>

    
    </tbody>
</table>

    

    
<dl class="details">
    

	

	

	

    

    

    

    

	

	

	

	

	



	

	

	

	
</dl>


    

    

    

    

    

    
    <h5>Returns:</h5>
    
            
<div class="param-desc">
    <p>C-term</p>
</div>




        

    
</dd>

        
            
<hr>
<dt>
    <h4 class="name" id="peptide_intern_mod"><span class="type-signature"></span>peptide_intern_mod(peptideSequence)</h4>
    
    
</dt>
<dd>

    
    <div class="description">
        <p>Function to analyze and return C-term mod of peptide sequence; in the sequence, modification is enclosed in ()</p>
    </div>
    

    

    

    
    
        <h5>Parameters:</h5>
        

<table class="params table table-striped">
    <thead>
    <tr>
        
        <th>Name</th>
        

        <th>Type</th>

        

        

        <th class="last">Description</th>
    </tr>
    </thead>

    <tbody>
    

        <tr>
            
                <td class="name"><code>peptideSequence</code></td>
            

            <td class="type">
            
                
<span class="param-type">string</span>



            
            </td>

            

            

            <td class="description last"><p>Input Protein one letter sequence</p></td>
        </tr>

    
    </tbody>
</table>

    

    
<dl class="details">
    

	

	

	

    

    

    

    

	

	

	

	

	



	

	

	

	
</dl>


    

    

    

    

    

    
    <h5>Returns:</h5>
    
            
<div class="param-desc">
    <p>internal-mod</p>
</div>




        

    
</dd>

        
            
<hr>
<dt>
    <h4 class="name" id="peptide_Nmod"><span class="type-signature"></span>peptide_Nmod(peptideSequence)</h4>
    
    
</dt>
<dd>

    
    <div class="description">
        <p>Function to analyze and return N-term  of peptide sequence; in the sequence, modification is enclosed in ()</p>
    </div>
    

    

    

    
    
        <h5>Parameters:</h5>
        

<table class="params table table-striped">
    <thead>
    <tr>
        
        <th>Name</th>
        

        <th>Type</th>

        

        

        <th class="last">Description</th>
    </tr>
    </thead>

    <tbody>
    

        <tr>
            
                <td class="name"><code>peptideSequence</code></td>
            

            <td class="type">
            
                
<span class="param-type">string</span>



            
            </td>

            

            

            <td class="description last"><p>Input Protein one letter sequence</p></td>
        </tr>

    
    </tbody>
</table>

    

    
<dl class="details">
    

	

	

	

    

    

    

    

	

	

	

	

	



	

	

	

	
</dl>


    

    

    

    

    

    
    <h5>Returns:</h5>
    
            
<div class="param-desc">
    <p>N-term</p>
</div>




        

    
</dd>

        
            
<hr>
<dt>
    <h4 class="name" id="protein_mw"><span class="type-signature"></span>protein_mw(proteinSequence)</h4>
    
    
</dt>
<dd>

    
    <div class="description">
        <p>Function to calculate MW (kDa) of protein sequence.<br>
<a href="http://www.bioinformatics.org/sms2/protein_mw.html">http://www.bioinformatics.org/sms2/protein_mw.html</a></p>
    </div>
    

    

    

    
    
        <h5>Parameters:</h5>
        

<table class="params table table-striped">
    <thead>
    <tr>
        
        <th>Name</th>
        

        <th>Type</th>

        

        

        <th class="last">Description</th>
    </tr>
    </thead>

    <tbody>
    

        <tr>
            
                <td class="name"><code>proteinSequence</code></td>
            

            <td class="type">
            
                
<span class="param-type">string</span>



            
            </td>

            

            

            <td class="description last"><p>Input Protein sequence</p></td>
        </tr>

    
    </tbody>
</table>

    

    
<dl class="details">
    

	

	

	

    

    

    

    

	

	

	

	

	



	

	

	

	
</dl>


    

    

    

    

    

    
    <h5>Returns:</h5>
    
            
<div class="param-desc">
    <p>MW (kDa) of protein sequence</p>
</div>




        

    
</dd>

        
            
<hr>
<dt>
    <h4 class="name" id="proteinGravy"><span class="type-signature"></span>proteinGravy(proteinSequence)</h4>
    
    
</dt>
<dd>

    
    <div class="description">
        <p>Function to calculate GRAVY (grand average of hydropathy) value for the protein sequences.<br>
<a href="http://www.bioinformatics.org/sms2/protein_gravy.html">http://www.bioinformatics.org/sms2/protein_gravy.html</a></p>
    </div>
    

    

    

    
    
        <h5>Parameters:</h5>
        

<table class="params table table-striped">
    <thead>
    <tr>
        
        <th>Name</th>
        

        <th>Type</th>

        

        

        <th class="last">Description</th>
    </tr>
    </thead>

    <tbody>
    

        <tr>
            
                <td class="name"><code>proteinSequence</code></td>
            

            <td class="type">
            
                
<span class="param-type">string</span>



            
            </td>

            

            

            <td class="description last"><p>Input Protein one letter sequence</p></td>
        </tr>

    
    </tbody>
</table>

    

    
<dl class="details">
    

	

	

	

    

    

    

    

	

	

	

	

	



	

	

	

	
</dl>


    

    

    

    

    

    
    <h5>Returns:</h5>
    
            
<div class="param-desc">
    <p>GRAVY of input protein</p>
</div>




        

    
</dd>

        
            
<hr>
<dt>
    <h4 class="name" id="proteinIep"><span class="type-signature"></span>proteinIep(proteinSequence, pKSet)</h4>
    
    
</dt>
<dd>

    
    <div class="description">
        <p>Function to calculate theoretical pI (isoelectric point) for the protein sequence.<br>
<a href="http://www.bioinformatics.org/sms2/protein_iep.html">http://www.bioinformatics.org/sms2/protein_iep.html</a></p>
    </div>
    

    

    

    
    
        <h5>Parameters:</h5>
        

<table class="params table table-striped">
    <thead>
    <tr>
        
        <th>Name</th>
        

        <th>Type</th>

        

        
        <th>Default</th>
        

        <th class="last">Description</th>
    </tr>
    </thead>

    <tbody>
    

        <tr>
            
                <td class="name"><code>proteinSequence</code></td>
            

            <td class="type">
            
                
<span class="param-type">string</span>



            
            </td>

            

            
                <td class="default">
                
                </td>
            

            <td class="description last"><p>Input Protein one letter sequence</p></td>
        </tr>

    

        <tr>
            
                <td class="name"><code>pKSet</code></td>
            

            <td class="type">
            
                
<span class="param-type">"DTAselect"</span>



            
            </td>

            

            
                <td class="default">
                
                    DTAselect
                
                </td>
            

            <td class="description last"><p>DTAselect (default) or EMBOSS</p></td>
        </tr>

    
    </tbody>
</table>

    

    
<dl class="details">
    

	

	

	

    

    

    

    

	

	

	

	

	



	

	

	

	
</dl>


    

    

    

    

    

    
    <h5>Returns:</h5>
    
            
<div class="param-desc">
    <p>theoretical pI of input protein</p>
</div>




        

    
</dd>

        
            
<hr>
<dt>
    <h4 class="name" id="proteinStats"><span class="type-signature"></span>proteinStats(proteinSequence)</h4>
    
    
</dt>
<dd>

    
    <div class="description">
        <p>Function to calculate protein stats.<br>
<a href="http://www.bioinformatics.org/sms2/protein_stats.html">http://www.bioinformatics.org/sms2/protein_stats.html</a></p>
    </div>
    

    

    

    
    
        <h5>Parameters:</h5>
        

<table class="params table table-striped">
    <thead>
    <tr>
        
        <th>Name</th>
        

        <th>Type</th>

        

        

        <th class="last">Description</th>
    </tr>
    </thead>

    <tbody>
    

        <tr>
            
                <td class="name"><code>proteinSequence</code></td>
            

            <td class="type">
            
                
<span class="param-type">string</span>



            
            </td>

            

            

            <td class="description last"><p>Input Protein one letter sequence</p></td>
        </tr>

    
    </tbody>
</table>

    

    
<dl class="details">
    

	

	

	

    

    

    

    

	

	

	

	

	



	

	

	

	
</dl>


    

    

    

    

    

    
    <h5>Returns:</h5>
    
            
<div class="param-desc">
    <p>AA composition stats of input protein</p>
</div>




        

    
</dd>

        
            
<hr>
<dt>
    <h4 class="name" id="restSummary"><span class="type-signature"></span>restSummary(dnaSequence, type)</h4>
    
    
</dt>
<dd>

    
    <div class="description">
        <p>Function to returns the number of commonly used restriction endonuclease cut sites of input DNA sequences.<br>
<a href="http://www.bioinformatics.org/sms2/rest_summary.html">http://www.bioinformatics.org/sms2/rest_summary.html</a></p>
    </div>
    

    

    

    
    
        <h5>Parameters:</h5>
        

<table class="params table table-striped">
    <thead>
    <tr>
        
        <th>Name</th>
        

        <th>Type</th>

        

        
        <th>Default</th>
        

        <th class="last">Description</th>
    </tr>
    </thead>

    <tbody>
    

        <tr>
            
                <td class="name"><code>dnaSequence</code></td>
            

            <td class="type">
            
                
<span class="param-type">string</span>



            
            </td>

            

            
                <td class="default">
                
                </td>
            

            <td class="description last"><p>Input DNA sequence</p></td>
        </tr>

    

        <tr>
            
                <td class="name"><code>type</code></td>
            

            <td class="type">
            
                
<span class="param-type">"circular"</span>



            
            </td>

            

            
                <td class="default">
                
                    circular
                
                </td>
            

            <td class="description last"><p>DNA sequence type: linear or circular (default)</p></td>
        </tr>

    
    </tbody>
</table>

    

    
<dl class="details">
    

	

	

	

    

    

    

    

	

	

	

	

	



	

	

	

	
</dl>


    

    

    

    

    

    
    <h5>Returns:</h5>
    
            
<div class="param-desc">
    <p>all common restriction sites in input sequence</p>
</div>




        

    
</dd>

        
            
<hr>
<dt>
    <h4 class="name" id="revComp"><span class="type-signature"></span>revComp(dnaSequence)</h4>
    
    
</dt>
<dd>

    
    <div class="description">
        <p>Function to reverse-complement counterpart of input DNA sequence.<br>
<a href="http://www.bioinformatics.org/sms2/rev_comp.html">http://www.bioinformatics.org/sms2/rev_comp.html</a></p>
    </div>
    

    

    

    
    
        <h5>Parameters:</h5>
        

<table class="params table table-striped">
    <thead>
    <tr>
        
        <th>Name</th>
        

        <th>Type</th>

        

        

        <th class="last">Description</th>
    </tr>
    </thead>

    <tbody>
    

        <tr>
            
                <td class="name"><code>dnaSequence</code></td>
            

            <td class="type">
            
                
<span class="param-type">string</span>



            
            </td>

            

            

            <td class="description last"><p>Input DNA sequence</p></td>
        </tr>

    
    </tbody>
</table>

    

    
<dl class="details">
    

	

	

	

    

    

    

    

	

	

	

	

	



	

	

	

	
</dl>


    

    

    

    

    

    
    <h5>Returns:</h5>
    
            
<div class="param-desc">
    <p>reverse-complement of input DNA sequence</p>
</div>




        

    
</dd>

        
            
<hr>
<dt>
    <h4 class="name" id="reverse"><span class="type-signature"></span>reverse(dnaSequence)</h4>
    
    
</dt>
<dd>

    
    <div class="description">
        <p>Function to reverse counterpart of input DNA sequence.<br>
<a href="http://www.bioinformatics.org/sms2/rev_comp.html">http://www.bioinformatics.org/sms2/rev_comp.html</a></p>
    </div>
    

    

    

    
    
        <h5>Parameters:</h5>
        

<table class="params table table-striped">
    <thead>
    <tr>
        
        <th>Name</th>
        

        <th>Type</th>

        

        

        <th class="last">Description</th>
    </tr>
    </thead>

    <tbody>
    

        <tr>
            
                <td class="name"><code>dnaSequence</code></td>
            

            <td class="type">
            
                
<span class="param-type">string</span>



            
            </td>

            

            

            <td class="description last"><p>Input DNA sequence</p></td>
        </tr>

    
    </tbody>
</table>

    

    
<dl class="details">
    

	

	

	

    

    

    

    

	

	

	

	

	



	

	

	

	
</dl>


    

    

    

    

    

    
    <h5>Returns:</h5>
    
            
<div class="param-desc">
    <p>reverse of input DNA sequence</p>
</div>




        

    
</dd>

        
            
<hr>
<dt>
    <h4 class="name" id="Seq1"><span class="type-signature"></span>Seq1(proteinSequence)</h4>
    
    
</dt>
<dd>

    
    <div class="description">
        <p>Function to convert three letters AA sequence to one letter.<br>
<a href="http://www.bioinformatics.org/sms2/three_to_one.html">http://www.bioinformatics.org/sms2/three_to_one.html</a></p>
    </div>
    

    

    

    
    
        <h5>Parameters:</h5>
        

<table class="params table table-striped">
    <thead>
    <tr>
        
        <th>Name</th>
        

        <th>Type</th>

        

        

        <th class="last">Description</th>
    </tr>
    </thead>

    <tbody>
    

        <tr>
            
                <td class="name"><code>proteinSequence</code></td>
            

            <td class="type">
            
                
<span class="param-type">string</span>



            
            </td>

            

            

            <td class="description last"><p>Input Protein three letter sequence</p></td>
        </tr>

    
    </tbody>
</table>

    

    
<dl class="details">
    

	

	

	

    

    

    

    

	

	

	

	

	



	

	

	

	
</dl>


    

    

    

    

    

    
    <h5>Returns:</h5>
    
            
<div class="param-desc">
    <p>one letter sequence of three letters sequence</p>
</div>




        

    
</dd>

        
            
<hr>
<dt>
    <h4 class="name" id="Seq3"><span class="type-signature"></span>Seq3(proteinSequence)</h4>
    
    
</dt>
<dd>

    
    <div class="description">
        <p>Function to convert one letter AA sequence to three letters</p>
    </div>
    

    

    

    
    
        <h5>Parameters:</h5>
        

<table class="params table table-striped">
    <thead>
    <tr>
        
        <th>Name</th>
        

        <th>Type</th>

        

        

        <th class="last">Description</th>
    </tr>
    </thead>

    <tbody>
    

        <tr>
            
                <td class="name"><code>proteinSequence</code></td>
            

            <td class="type">
            
                
<span class="param-type">string</span>



            
            </td>

            

            

            <td class="description last"><p>Input Protein one letter sequence</p></td>
        </tr>

    
    </tbody>
</table>

    

    
<dl class="details">
    

	

	

	

    

    

    

    

	

	

	

	

	



	

	

	

	
</dl>


    

    

    

    

    

    
    <h5>Returns:</h5>
    
            
<div class="param-desc">
    <p>three letter sequence of one letter sequence</p>
</div>




        

    
</dd>

        
            
<hr>
<dt>
    <h4 class="name" id="SeqDiff"><span class="type-signature"></span>SeqDiff(SequenceOne, SequenceTwo)</h4>
    
    
</dt>
<dd>

    
    <div class="description">
        <p>Function to compare two sequences (dna/protein) with same length and return their differences</p>
    </div>
    

    

    

    
    
        <h5>Parameters:</h5>
        

<table class="params table table-striped">
    <thead>
    <tr>
        
        <th>Name</th>
        

        <th>Type</th>

        

        

        <th class="last">Description</th>
    </tr>
    </thead>

    <tbody>
    

        <tr>
            
                <td class="name"><code>SequenceOne</code></td>
            

            <td class="type">
            
                
<span class="param-type">string</span>



            
            </td>

            

            

            <td class="description last"><p>Input one letter sequence 1</p></td>
        </tr>

    

        <tr>
            
                <td class="name"><code>SequenceTwo</code></td>
            

            <td class="type">
            
                
<span class="param-type">string</span>



            
            </td>

            

            

            <td class="description last"><p>Input one letter sequence 2</p></td>
        </tr>

    
    </tbody>
</table>

    

    
<dl class="details">
    

	

	

	

    

    

    

    

	

	

	

	

	



	

	

	

	
</dl>


    

    

    

    

    

    
    <h5>Returns:</h5>
    
            
<div class="param-desc">
    <p>sequence differences</p>
</div>




        

    
</dd>

        
            
<hr>
<dt>
    <h4 class="name" id="translate"><span class="type-signature"></span>translate(dnaSequence, type)</h4>
    
    
</dt>
<dd>

    
    <div class="description">
        <p>Function to converts DNA into a protein in the reading frame.<br>
<a href="http://www.bioinformatics.org/sms2/translate.html">http://www.bioinformatics.org/sms2/translate.html</a></p>
    </div>
    

    

    

    
    
        <h5>Parameters:</h5>
        

<table class="params table table-striped">
    <thead>
    <tr>
        
        <th>Name</th>
        

        <th>Type</th>

        

        
        <th>Default</th>
        

        <th class="last">Description</th>
    </tr>
    </thead>

    <tbody>
    

        <tr>
            
                <td class="name"><code>dnaSequence</code></td>
            

            <td class="type">
            
                
<span class="param-type">string</span>



            
            </td>

            

            
                <td class="default">
                
                </td>
            

            <td class="description last"><p>Input DNA sequence</p></td>
        </tr>

    

        <tr>
            
                <td class="name"><code>type</code></td>
            

            <td class="type">
            
                
<span class="param-type">"standard"</span>



            
            </td>

            

            
                <td class="default">
                
                    standard
                
                </td>
            

            <td class="description last"><p>standard (default) or transl_table=1 etc see <a href="http://www.bioinformatics.org/sms2/genetic_code.html">http://www.bioinformatics.org/sms2/genetic_code.html</a></p></td>
        </tr>

    
    </tbody>
</table>

    

    
<dl class="details">
    

	

	

	

    

    

    

    

	

	

	

	

	



	

	

	

	
</dl>


    

    

    

    

    

    
    <h5>Returns:</h5>
    
            
<div class="param-desc">
    <p>translated DNA sequence</p>
</div>




        

    
</dd>

        </dl>
    

    

    
</article>

</section>




		</div>
	</div>

	<div class="clearfix"></div>

	
		<div class="col-md-3">
			<div id="toc" class="col-md-3 hidden-xs hidden-sm hidden-md"></div>
		</div>
	

</div>
</div>



<footer>


	<span class="copyright">
	DocStrap Copyright Â© 2012-2015 The contributors to the JSDoc3 and DocStrap projects.
	</span>

<span class="jsdoc-message">
	Documentation generated by <a href="https://github.com/jsdoc3/jsdoc">JSDoc 3.6.4</a>
	
	using the <a href="https://github.com/docstrap/docstrap">DocStrap template</a>.
</span>
</footer>

<script src="scripts/docstrap.lib.js"></script>
<script src="scripts/toc.js"></script>


<script>
$( function () {
	$( "[id*='$']" ).each( function () {
		var $this = $( this );

		$this.attr( "id", $this.attr( "id" ).replace( "$", "__" ) );
	} );

	$( ".tutorial-section pre, .readme-section pre, pre.prettyprint.source" ).each( function () {
		var $this = $( this );

		var example = $this.find( "code" );
		exampleText = example.html();
		var lang = /{@lang (.*?)}/.exec( exampleText );
		if ( lang && lang[1] ) {
			exampleText = exampleText.replace( lang[0], "" );
			example.html( exampleText );
			lang = lang[1];
		} else {
			var langClassMatch = example.parent()[0].className.match(/lang\-(\S+)/);
			lang = langClassMatch ? langClassMatch[1] : "javascript";
		}

		if ( lang ) {

			$this
			.addClass( "sunlight-highlight-" + lang )
			.addClass( "linenums" )
			.html( example.html() );

		}
	} );

	Sunlight.highlightAll( {
		lineNumbers : true,
		showMenu : true,
		enableDoclinks : true
	} );

	$.catchAnchorLinks( {
        navbarOffset: 10
	} );
	$( "#toc" ).toc( {
		anchorName  : function ( i, heading, prefix ) {
			return $( heading ).attr( "id" ) || ( prefix + i );
		},
		selectors   : "#toc-content h1,#toc-content h2,#toc-content h3,#toc-content h4",
		showAndHide : false,
		smoothScrolling: true
	} );

	$( "#main span[id^='toc']" ).addClass( "toc-shim" );
	$( '.dropdown-toggle' ).dropdown();

    $( "table" ).each( function () {
      var $this = $( this );
      $this.addClass('table');
    } );

} );
</script>



<!--Navigation and Symbol Display-->


<!--Google Analytics-->




</body>
</html>